### PR TITLE
Unlock platinum trophies

### PR DESCRIPTION
**Pull request by Megamouse**
This should unlock platinum trophies once you unlock the last necessary regular trophy.
Please back up your save and trophy dirs before testing.

### CODE REVIEW

**Comment by isshininu** [comment]
Holy shit

https://github.com/RPCS3/rpcs3/issues/4289 this will close it

File: rpcs3/Loader/TROPUSR.cpp [hunk 1]
```diff
@@ -215,6 +216,46 @@ u32 TROPUSRLoader::GetUnlockedTrophiesCount()
...
+			if (trophy_info.trophy_id == trophy.trophy_id)
+			{
+				// The platinum trophy stays locked if any relevant trophy is still locked
+				if (trophy_info.trophy_pid == pid && !trophy.trophy_state)
+				{
+					return invalid_trophy_id;
+				}
+
+				// Onto the next trophy id
+				break;
```
Comment (CookiePLMonster) [inline rpcs3/Loader/TROPUSR.cpp]: Unless I'm misunderstanding the intent, this begs for a `std::find_if`,

**Comment by Megamouse** [comment]
> Holy shit
> 
> #4289 this will close it
> [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3964289/RPCS3.log.gz)
> Edit: not working on DeS platinum

Can you send me the trophy file that comes with the game?

File: rpcs3/Loader/TROPUSR.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Loader/TROPUSR.cpp]: Yeah I was just too lazy because I had no idea if this would work anyway

**Comment by isshininu** [comment]
This?
[TROPDIR.zip](https://github.com/RPCS3/rpcs3/files/3964314/TROPDIR.zip)

**Comment by Megamouse** [comment]
I probably have to unlock the trophy myself instead of only telling the game that the requirements were met. Needs some more investigation.

**Comment by Megamouse** [comment]
It should work now

**Comment by isshininu** [comment]
On game load, nothing changes. Unfortunately I forgot to make a backup with 2 achievements left, so here is only platinum remains unlocked.

**Comment by Megamouse** [comment]
I guess we'll never know since you seem to be the only tester

**Comment by Megamouse** [comment]
In case you still have the savegame I can revert your trophy states. Maybe evenone that's easier to achieve

**Comment by isshininu** [comment]
What exactly you want me to do?
If you want to edit out one of the achievements, any of the full upgrade weapon trophies are looks most easy to revert, as I can try to add it immediately again.

**Comment by Megamouse** [comment]
The one in home/trophystuff/gameid

**Comment by isshininu** [comment]
[NPWR01249_00.zip](https://github.com/RPCS3/rpcs3/files/3975389/NPWR01249_00.zip)
Sure, here it is.

**Comment by Megamouse** [comment]
@isshininu I reverted the last change because we have some possibly related changes in master now.
I added some hacky code that lets you lock/unlock trophies by double click in the trophy manager while the game is stopped.

Could you please backup your savegames and trophies and try to unlock one of the last trophies (**not premium**)?

If this doesn't work then I will add my newer code again and we can hopefully make this work.

**Comment by isshininu** [comment]
It looks like locking trophy in manager resets right after closing trophy manager. When opening it again, all locked trophy are already unlocked.

**Comment by Megamouse** [comment]
@isshininu lol... Yeah I made a mistake. Should work in the next build

**Comment by GitArUs** [comment]
Tested on unnamed game. Locking selected trophy by double click worked well. I have unlocked it again by performing appropriate action in game, but platinum remains locked.

**Comment by Megamouse** [comment]
Can you guys please upload logs when after testing. I can't really deduce anything with that information alone.

**Comment by GitArUs** [comment]
Here You are. You will see two unlocks there, the second one failed, because already_unlocked.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3980336/RPCS3.log.gz)

**Comment by Megamouse** [comment]
can you please upload the trophy folder "dev_hdd0\home\00000001\trophy\NPWR11866_00"

**Comment by GitArUs** [comment]
I hope You don't need all png icons ?...
[tropdir.tar.gz](https://github.com/RPCS3/rpcs3/files/3980386/tropdir.tar.gz)

**Comment by isshininu** [comment]
Here is my log after reverting one of the bronze trophys to locked state and trying to achieve it again. Bronze achieved, platinum is not.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3989699/RPCS3.log.gz)

**Comment by Megamouse** [comment]
Okay, let's try this by unlocking it ourself, since the game obviously didn't care to do it.

**Comment by isshininu** [comment]
Unlocking platinum?

**Comment by Megamouse** [comment]
no. I mean just try it again as before.

**Comment by isshininu** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3990018/RPCS3.log.gz)
notning

**Comment by Megamouse** [comment]
please also send the trophies again

**Comment by GitArUs** [comment]
Doesn't work for me either.
[RPCS3_2.log.gz](https://github.com/RPCS3/rpcs3/files/3990606/RPCS3_2.log.gz)
[tropdir_2.tar.gz](https://github.com/RPCS3/rpcs3/files/3990608/tropdir_2.tar.gz)

**Comment by Megamouse** [comment]
Allright. I noticed that a specific value was missing because old trophy files do not contain it yet.
This should now work even with old data.

I also noticed that I was restricting the unlock mechanism to a certain condition, which was illogical (but didn't cause the problem you guys are having).

**Comment by isshininu** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3990988/RPCS3.log.gz)
Yeah,  it works. No platinum notification in game, but after unlocking bronze platinum WAS here in trophy manager..

**Comment by GitArUs** [comment]
Confirmed. Platinum unlocked, but no trophy popup (maybe hidden under first popup ?).
[RPCS3_3.log.gz](https://github.com/RPCS3/rpcs3/files/3991202/RPCS3_3.log.gz)
[tropdir_3.tar.gz](https://github.com/RPCS3/rpcs3/files/3991203/tropdir_3.tar.gz)

**Comment by Megamouse** [comment]
Yes, showing the trophies queued or at the same time isn't implemented yet

**Comment by kd-11** [comment]
Queued popups implemented with https://github.com/RPCS3/rpcs3/pull/7186. If building locally, you can check if combining the 2 PRs makes popups work, otherwise just wait a bit for it to be merged.

**Comment by GitArUs** [comment]
I'm not sure if I did it right (merged one commit from #7184 and two commits from #7186 to this PR). Popup animations work (slide-in and slide-out), but still only one trophy popup is visible, platinum trophy unlocks but there is no platinum popup.

**Comment by Megamouse** [comment]
I haven't applied the new logic yet, so there's no way it could've worked yet.

**Comment by Megamouse** [comment]
You can try again now.

**Comment by Megamouse** [comment]
@isshininu can you please test again?

**Comment by isshininu** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/4023713/RPCS3.log.gz)
Снимок экрана (1012)
Снимок экрана (1013)

**Comment by kd-11** [comment]
Nice

**Comment by GitArUs** [comment]
That was fast. Thanks. Works as expected.

**Pull request merged**